library/std/src/io/buffered/bufwriter.rs

```diff
@@ -291,64 +374,145 @@ impl<W: Write> BufWriter<W> {
-            self.panicked = false;
-            r
+        if let Some(buf) = only_one(bufs, |b| !b.is_empty()) {
+            // If there's exactly 1 incoming buffer, `Self::write` can make
+            // use of self.inner.write_vectored to attempt to combine flushing
+            // the existing buffer with writing the new one.
+            self.write(buf)
+        } else if self.get_ref().is_write_vectored() {
+            let total_len: usize = bufs.iter().map(|buf| buf.len()).sum();
+
+            if total_len > self.available() {
+                self.flush_buf()?;
+            }
+
+            if total_len >= self.available() {
+                self.get_mut().write_vectored(bufs)
+            } else {
+                // Correctness note: we've already verified that none of these
+                // will overflow the buffer, because total_len < capacity
+                bufs.iter().for_each(|buf| self.buf.extend_from_slice(buf));
+                Ok(total_len)
+            }
         } else {
-            bufs.iter().for_each(|b| self.buf.extend_from_slice(b));
-            Ok(total_len)
+            // Because the inner writer doesn't have native vectored write
+            // support, we should take care of buffering together the individual
+            // incoming bufs, even if the *total* length is larger than our
+            // buffer. We only want to skip our buffer if an *individual* write
+            // exceeds our buffer capacity.
+            let mut total_buffered = 0;
+
+            for buf in bufs {
+                if total_buffered == 0 {
+                    if buf.len() > self.available() {
+                        // If an individual write would overflow our remaining
+                        // capacity and we haven't buffered anything yet,
+                        // pre-flush before buffering (same as with regular
+                        // write()).
+                        self.flush_buf()?;
+                    }
+
+                    if buf.len() >= self.available() {
+                        // If an individual buffer exceeds our *total* capacity
+                        // and we haven't buffered anything yet, just forward
+                        // it to the underlying device
+                        return self.get_mut().write(buf);
+                    }
+                }
+
+                // Buffer as much as possible until we reach full capacity.
+                // Once we've buffered at least 1 byte, we're obligated to
+                // return Ok(..) before attempting any fallible i/o operations,
+                // so once the buffer is full we immediately return.
+                total_buffered += self.write_to_buf(buf);
+                if self.available() == 0 {
+                    break;
+                }
+            }
+
+            Ok(total_buffered)
         }
     }
 
     fn is_write_vectored(&self) -> bool {
-        self.get_ref().is_write_vectored()
+        true
```

Comment:
Is that really an improvement? It would result in the caller going through the effort of managing IoSlices only for BufWriter to disassemble them into the buffer, assuming they're consisting of small writes. And in case of large writes we would only try to pass through the first one.

The caller might as well call `write()` multiple times to have things aggregate into the buffer.

Reply:
I'm certainly assuming that no one is *going out of their way* to use vectored writes if it isn't imminently convenient to do so, and in particular they probably aren't dynamically doing their own "buffering" of IoSlice objects to eventually forward. For instance, `write_fmt` isn't going to be creating a buffer of IoSlice for writing the constitutent parts. I'm instead assuming that people are using vectored writes in cases where they happen to already have a fixed set of buffers available; for instance, an HTTP protocol might use it to forward HTTP headers, or a variant of `println` might use it to append a newline to the base content.

This reflects my own usage of `write_vectored` in `BufWriter`; I only call it in cases where I conveniently have a fixed set of buffers I'm interested in writing.